### PR TITLE
OHRM5X-186 : Admin - Nationalities

### PR DESCRIPTION
**Pull request by ireshe**
added the Nationality FE & BE

### CODE REVIEW

File: symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     * @throws Exception
+     */
+    public function create(): EndpointCreateResult
+    {
+        // TODO:: Check data group permission
+        $nationalities = $this->saveNationality();
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php:141]: `$nationalities` ==> `$nationality`

File: symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+     * @throws DaoException
+     * @throws RecordNotFoundException
+     */
+    public function saveNationality(): Nationality
+    {
+        $id = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, CommonParams::PARAMETER_ID);
+        $name = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_NAME);
+        if (!empty($id)) {
+            $nationality = $this->getNationalityService()->getNationalityById($id);
+            if ($nationality == null) {
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php:157]: use `$this->throwRecordNotFoundExceptionIfNotExist`

File: symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+     * @inheritDoc
+     * @throws RecordNotFoundException
+     * @throws Exception
+     */
+    public function getOne(): EndpointGetOneResult
+    {
+        // TODO:: Check data group permission
+        $id = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, CommonParams::PARAMETER_ID);
+        $nationality = $this->getNationalityService()->getNationalityById($id);
+        if (!$nationality instanceof Nationality) {
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php:87]: use `$this->throwRecordNotFoundExceptionIfNotExist`

File: symfony/plugins/orangehrmAdminPlugin/Dao/NationalityDao.php [hunk 4]
```diff
@@ -57,26 +99,66 @@ public function getNationalityById(int $id): ?Nationality
...
+            throw new DaoException($e->getMessage(), $e->getCode(), $e);
+        }
+    }
+
+    /**
+     * @param string $NationalityName
+     * @return bool
+     * @throws DaoException
+     */
+    public function isExistingNationalityName(string $NationalityName): bool
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/NationalityDao.php:148]: refactor rename `$NationalityName` => `$nationalityName`, update doc comment as well

File: symfony/plugins/orangehrmAdminPlugin/Dao/NationalityDao.php [hunk 5]
```diff
@@ -57,26 +99,66 @@ public function getNationalityById(int $id): ?Nationality
...
+    {
+        try {
+            $q = $this->createQueryBuilder(Nationality::class, 'n');
+            $trimmed = trim($NationalityName, ' ');
+            $q->Where('n.name = :name');
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/NationalityDao.php:153]: `$q->where('n.name = :name');`

File: symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+use OrangeHRM\Core\Api\V2\Serializer\EndpointGetOneResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointUpdateResult;
+use OrangeHRM\Core\Api\V2\Validator\ParamRule;
+use OrangeHRM\Core\Api\V2\Validator\ParamRuleCollection;
+use OrangeHRM\Core\Api\V2\Validator\Rule;
+use OrangeHRM\Core\Api\V2\Validator\Rules;
+use OrangeHRM\Core\Exception\DaoException;
+use OrangeHRM\Entity\Nationality;
+
+class NationalityAPI extends EndPoint implements CrudEndpoint
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php:45]: write unit tests for API, refer `symfony/plugins/orangehrmPimPlugin/test/Api/EmploymentContractAPITest.php`

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/client/src/orangehrmAdminPlugin/pages/nationality/EditNationality.vue [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+      .then(response => {
+        const {data} = response.data;
+        this.nationality.id = data.id;
+        this.nationality.name = data.name;
+        // Fetch list data for unique test
+        return this.http.getAll();
+      })
+      .then(response => {
+        const {data} = response.data;
+        this.rules.name.push(v => {
```
Comment (RajithaKumara) [inline symfony/client/src/orangehrmAdminPlugin/pages/nationality/EditNationality.vue:122]: import
```js
import {
  required,
  shouldNotExceedCharLength,
} from '@orangehrm/core/util/validation/rules';
```

Refer `symfony/client/src/orangehrmPimPlugin/pages/employee/SaveEmployee.vue` validations

File: symfony/client/src/orangehrmAdminPlugin/pages/nationality/EditNationality.vue [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+          return (!!v && v.trim() !== '') || 'Required';
+        });
+        this.rules.name.push(v => {
```
Comment (RajithaKumara) [inline symfony/client/src/orangehrmAdminPlugin/pages/nationality/EditNationality.vue:125]: replace rule with `shouldNotExceedCharLength(100)`

File: symfony/client/src/orangehrmAdminPlugin/pages/nationality/SaveNationality.vue [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    },
+  },
+
+  created() {
+    this.isLoading = true;
+    this.http
+      .getAll()
+      .then(response => {
+        const {data} = response.data;
+        this.rules.name.push(v => {
```
Comment (RajithaKumara) [inline symfony/client/src/orangehrmAdminPlugin/pages/nationality/SaveNationality.vue:110]: update rules as mention in `EditNationality.vue` review comment

File: symfony/plugins/orangehrmAdminPlugin/Api/EducationAPI.php [hunk 10]
```diff
@@ -191,6 +187,7 @@ public function update(): EndpointResourceResult
     public function getValidationRuleForUpdate(): ParamRuleCollection
     {
         return new ParamRuleCollection(
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/EducationAPI.php:190]: unnecessary new line

File: symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php [hunk 11]
```diff
@@ -194,8 +191,8 @@ public function update(): EndpointResourceResult
     public function getValidationRuleForUpdate(): ParamRuleCollection
     {
         return new ParamRuleCollection(
-            new ParamRule(
-                CommonParams::PARAMETER_ID,
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:194]: unnecessary new line

File: symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    /**
+     * @return NationalityService
+     */
+    public function getNationalityService(): NationalityService
+    {
+        if (!$this->nationalityService instanceof NationalityService) {
+            $this->nationalityService = new NationalityService();
+        }
+        return $this->nationalityService;
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php:61]: unnecessary new line

File: symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    }
+
+    /**
+     * @param NationalityService|null $nationalityService
+     */
+    public function setNationalityService(?NationalityService $nationalityService): void
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php:67]: remove if not used

File: symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+    {
+        $this->nationalityService = $nationalityService;
+    }
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/NationalityAPI.php:71]: multiple unnecessary blank lines

File: symfony/plugins/orangehrmAdminPlugin/test/Api/NationalityAPITest.php [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+ * OrangeHRM is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Admin\test\Api;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Api/NationalityAPITest.php:20]: ==> `namespace OrangeHRM\Tests\Admin\Api;`

File: symfony/plugins/orangehrmPimPlugin/Controller/EmployeePersonalDetailController.php [hunk 16]
```diff
@@ -25,6 +25,7 @@
 use OrangeHRM\Core\Vue\Prop;
 use OrangeHRM\Framework\Http\Request;
 
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Controller/EmployeePersonalDetailController.php:28]: unnecessary new line

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/client/src/orangehrmAdminPlugin/pages/nationality/EditNationality.vue [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+      this.http
+        .update(this.nationalityId, {
+          name: this.nationality.name,
+        })
+        .then(() => {
+          return this.$toast.updateSuccess();
+        })
+        .then(() => {
+          this.onCancel();
+          this.isLoading = false;
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/nationality/EditNationality.vue:105]: remove this. let redirect back while loading.

File: symfony/client/src/orangehrmAdminPlugin/pages/nationality/SaveNationality.vue [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+  methods: {
+    onSave() {
+      this.isLoading = true;
+      this.http
+        .create({
+          name: this.nationality.name,
+        })
+        .then(() => {
+          return this.$toast.addSuccess();
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/nationality/SaveNationality.vue:91]: use `return this.$toast.saveSuccess()`

File: symfony/client/src/orangehrmAdminPlugin/pages/nationality/SaveNationality.vue [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+        })
+        .then(() => {
+          this.nationality.name = '';
+          this.isLoading = false;
+          this.onCancel();
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/nationality/SaveNationality.vue:96]: just call `this.onCancel();` do not end loading.

Comment (ireshe) [reply symfony/client/src/orangehrmAdminPlugin/pages/nationality/SaveNationality.vue:98]: I removed isloading and nationality.name

**Pull request merged**